api/handlers.go

```diff
@@ -161,6 +172,169 @@ func UnpackHandler(routePrefix string) http.HandlerFunc {
+	p.NewName = util.ReqParamBool(r, "newName", false)
+	dsBytes := []byte(r.FormValue("dataset"))
+	if len(dsBytes) != 0 {
+		p.Dataset = &dataset.Dataset{}
+		err := p.Dataset.UnmarshalJSON(dsBytes)
+		if err != nil {
+			return err
+		}
+	}
+	log.Debugw("parseSaveParamsFromRequest", "params", p)
+	return nil
+}
+
+// formFileDataset extracts a dataset document from a http Request
+func formFileDataset(r *http.Request, ds *dataset.Dataset) (err error) {
+	datafile, dataHeader, err := r.FormFile("file")
+	if err == http.ErrMissingFile {
+		err = nil
+	} else if err != nil {
+		err = fmt.Errorf("error opening dataset file: %s", err)
+		return
+	}
+	if datafile != nil {
+		switch strings.ToLower(filepath.Ext(dataHeader.Filename)) {
+		case ".yaml", ".yml":
+			var data []byte
+			data, err = ioutil.ReadAll(datafile)
+			if err != nil {
+				err = fmt.Errorf("reading dataset file: %w", err)
+				return
+			}
+			fields := &map[string]interface{}{}
+			if err = yaml.Unmarshal(data, fields); err != nil {
+				err = fmt.Errorf("deserializing YAML file: %w", err)
+				return
+			}
+			if err = fill.Struct(*fields, ds); err != nil {
+				return
+			}
+		case ".json":
+			if err = json.NewDecoder(datafile).Decode(ds); err != nil {
+				err = fmt.Errorf("error decoding json file: %w", err)
+				return
+			}
+		}
+	}
+
+	if peername := r.FormValue("peername"); peername != "" {
+		ds.Peername = peername
+	}
+	if name := r.FormValue("name"); name != "" {
+		ds.Name = name
+	}
+	if bp := r.FormValue("body_path"); bp != "" {
+		ds.BodyPath = bp
+	}
+
+	tfFile, tfHeader, err := r.FormFile("transform")
+	if err == http.ErrMissingFile {
+		err = nil
+	} else if err != nil {
+		err = fmt.Errorf("error opening transform file: %w", err)
+		return
+	}
+	if tfFile != nil {
+		if ds.Transform == nil {
+			ds.Transform = &dataset.Transform{}
+		}
+		ds.Transform.SetScriptFile(qfs.NewMemfileReader(tfHeader.Filename, tfFile))
+	}
+
+	vizFile, vizHeader, err := r.FormFile("viz")
+	if err == http.ErrMissingFile {
+		err = nil
+	} else if err != nil {
+		err = fmt.Errorf("error opening viz file: %w", err)
+		return
+	}
+	if vizFile != nil {
+		if ds.Viz == nil {
+			ds.Viz = &dataset.Viz{}
+		}
+		ds.Viz.SetScriptFile(qfs.NewMemfileReader(vizHeader.Filename, vizFile))
+	}
+
+	readmeFile, readmeHeader, err := r.FormFile("readme")
+	if err == http.ErrMissingFile {
+		err = nil
+	} else if err != nil {
+		err = fmt.Errorf("error opening readme file: %w", err)
+		return
+	}
+	if readmeFile != nil {
+		if ds.Readme == nil {
+			ds.Readme = &dataset.Readme{}
+			ds.Readme.SetScriptFile(qfs.NewMemfileReader(readmeHeader.Filename, readmeFile))
+		}
+	}
+
+	bodyfile, bodyHeader, err := r.FormFile("body")
+	if err == http.ErrMissingFile {
+		err = nil
+	} else if err != nil {
+		err = fmt.Errorf("error opening body file: %w", err)
+		return
+	}
+	if bodyfile != nil {
+		ds.BodyPath = bodyHeader.Filename
+		ds.SetBodyFile(qfs.NewMemfileReader(bodyHeader.Filename, bodyfile))
```

Comment:
This is the biggest departure from previous iterations:

Rather than opening the file & consuming the reader to inline the bytes, we instead just attach the reader as a script file to the body (or viz, or transform, or readme)